src/ws/ws_manager.rs

```diff
@@ -166,32 +170,74 @@ impl WsManager {
             }
             Message::Notification(_) => Ok("notification".to_string()),
             Message::SubscriptionResponse | Message::Pong => Ok(String::default()),
+            Message::NoData => Ok(String::default()),
+            Message::HyperliquidError(err) => Ok(format!("hyperliquid error: {err:?}")),
         }
     }
 
     async fn parse_and_send_data(
         data: Option<std::result::Result<protocol::Message, tungstenite::Error>>,
         subscriptions: &Arc<Mutex<HashMap<String, Vec<SubscriptionData>>>>,
     ) -> Result<()> {
-        let data = data
-            .ok_or(Error::ReaderDataNotFound)?
-            .map_err(|e| Error::GenericReader(e.to_string()))?
-            .into_text()
-            .map_err(|e| Error::ReaderTextConversion(e.to_string()))?;
-        if !data.starts_with('{') {
-            return Ok(());
-        }
-        let message =
-            serde_json::from_str::<Message>(&data).map_err(|e| Error::JsonParse(e.to_string()))?;
-        let identifier = WsManager::get_identifier(&message)?;
-        if identifier.is_empty() {
-            return Ok(());
+        match data {
+            Some(data) => match data {
```

Comment:
This level of nesting makes it very hard to read. Where there is an option, can you use the pattern `let Some(variable_name) = .. else { early return };`?

Reply:
Changed. The NoData condition can be written that way.